packages/ra-core/src/controller/input/useReferenceArrayInputController.ts

```diff
@@ -0,0 +1,145 @@
+import { useMemo, useState } from 'react';
+
+import { Record, Pagination, Sort } from '../../types';
+import { useGetMany } from '../../dataProvider';
+import { FieldInputProps } from 'react-final-form';
+import useGetMatching from '../../dataProvider/useGetMatching';
+import { useTranslate } from '../../i18n';
+import { getStatusForArrayInput as getDataStatus } from './referenceDataStatus';
+
+/**
+ * @typedef ReferenceArrayProps
+ * @type {Object}
+ * @property {Array} ids the list of ids.
+ * @property {Object} data Object holding the reference data by their ids
+ * @property {Object} error the error returned by the dataProvider
+ * @property {boolean} loading is the reference currently loading
+ * @property {boolean} loaded has the reference already been loaded
+ * @property {string} referenceBasePath basePath of the reference
+ */
+interface ReferenceArrayInputProps {
+    choices: Record[];
+    error?: any;
+    warning?: any;
+    loading: boolean;
+    loaded: boolean;
+    referenceBasePath: string;
+    setFilter: (filter: any) => void;
+    setPagination: (pagination: Pagination) => void;
+    setSort: (sort: Sort) => void;
+}
+
+interface Option {
+    basePath: string;
+    filter?: any;
+    filterToQuery?: (filter: any) => any;
+    input: FieldInputProps<any, HTMLElement>;
+    options?: any;
+    perPage?: number;
+    record?: Record;
+    reference: string;
+    resource: string;
+    sort?: Sort;
+    source: string;
+}
+
+/**
+ * Prepare data for the ReferenceArrayInput components
+ *
+ * @example
+ *
+ * const { choices, error, loaded, loading, referenceBasePath } = useReferenceArrayInputController({
+ *      basePath: 'resource';
+ *      record: { referenceIds: ['id1', 'id2']};
+ *      reference: 'reference';
+ *      resource: 'resource';
+ *      source: 'referenceIds';
+ * });
+ *
+ * @param {Object} option
+ * @param {boolean} option.allowEmpty do we allow for no referenced record (default to false)
+ * @param {string} option.basePath basepath to current resource
+ * @param {string | false} option.linkType The type of the link toward the referenced record. edit, show of false for no link (default to edit)
+ * @param {Object} option.record The The current resource record
+ * @param {string} option.reference The linked resource name
+ * @param {string} option.resource The current resource name
+ * @param {string} option.source The key of the linked resource identifier
+ *
+ * @return {Object} controllerProps Fetched data and callbacks for the ReferenceArrayInput components
+ */
+const useReferenceArrayInputController = ({
+    basePath,
+    filter: defaultFilter,
+    filterToQuery = searchText => ({ q: searchText }),
```

Comment:
Move the default value to a constant to avoid having a new function at every render, defeating your useMemo line 93